contracts/extensions/ExpenditureUtils.sol

```diff
@@ -0,0 +1,318 @@
+
+    colony = IColony(_colony);
+  }
+
+  /// @notice Called when upgrading the extension
+  function finishUpgrade() public override auth {}
+
+  /// @notice Called when deprecating (or undeprecating) the extension
+  function deprecate(bool _deprecated) public override auth {
+    deprecated = _deprecated;
+  }
+
+  /// @notice Called when uninstalling the extension
+  function uninstall() public override auth {
+    selfdestruct(address(uint160(address(colony))));
+  }
+
+  // Public
+
+  function setStakeFraction(uint256 _stakeFraction) public onlyRoot {
+    require(_stakeFraction <= WAD, "expenditure-utils-value-too-large");
+    stakeFraction = _stakeFraction;
+  }
+
+  function makeExpenditureWithStake(
+    uint256 _permissionDomainId,
+    uint256 _childSkillIndex,
+    uint256 _domainId,
+    bytes memory _key,
+    bytes memory _value,
+    uint256 _branchMask,
+    bytes32[] memory _siblings
+  )
+    public
+  {
+    uint256 domainRep = getReputationFromProof(_domainId, _key, _value, _branchMask, _siblings);
+    uint256 stakeAmount = wmul(domainRep, stakeFraction);
+
+    colony.obligateStake(msgSender(), _domainId, stakeAmount);
+    uint256 expenditureId = colony.makeExpenditure(_permissionDomainId, _childSkillIndex, _domainId);
+
+    stakes[expenditureId] = Stake({ creator: msgSender(), amount: stakeAmount });
+    colony.transferExpenditure(expenditureId, msgSender());
+
+    emit ExpenditureMadeViaStake(msgSender(), expenditureId, stakeAmount);
+  }
+
+  function reclaimStake(uint256 _expenditureId) public {
+    Stake storage stake = stakes[_expenditureId];
+    require(stake.amount > 0, "expenditure-utils-nothing-to-claim");
+
+    ColonyDataTypes.Expenditure memory expenditure = colony.getExpenditure(_expenditureId);
+    require(
+      expenditure.status == ColonyDataTypes.ExpenditureStatus.Cancelled ||
+      expenditure.status == ColonyDataTypes.ExpenditureStatus.Finalized,
+      "expenditure-utils-expenditure-invalid-state"
+    );
+
+    colony.deobligateStake(stake.creator, expenditure.domainId, stake.amount);
+
+  // slither-disable-next-line reentrancy-no-eth
+    delete stakes[_expenditureId];
+  }
+
+  function cancelAndReclaimStake(
+    uint256 _permissionDomainId,
+    uint256 _childSkillIndex,
+    uint256 _expenditureId
+  )
+    public
+  {
+    Stake storage stake = stakes[_expenditureId];
+    ColonyDataTypes.Expenditure memory expenditure = colony.getExpenditure(_expenditureId);
+
+    require(expenditure.owner == msgSender(), "expenditure-utils-must-be-owner");
+
+    require(
+      expenditure.status == ColonyDataTypes.ExpenditureStatus.Draft,
+      "expenditure-utils-expenditure-not-draft"
+    );
+
+    cancelExpenditure(_permissionDomainId, _childSkillIndex, _expenditureId, expenditure.owner);
+    reclaimStake(_expenditureId);
+  }
+
+  function cancelAndPunish(
+    uint256 _permissionDomainId,
+    uint256 _childSkillIndex,
```

Comment:
This function only works if the caller and the extension have the same permission proof, which is not necessarily the case?